win32/tests/file_win32_ut/file_win32_ut.c

```diff
@@ -0,0 +1,952 @@
+            ASSERT_IS_NULL(file_handle);
+        }
+    }
+}
+
+/*Tests_SRS_FILE_43_005: [ If handle is NULL, file_destroy shall return. ]*/
+/*Tests_SRS_FILE_WIN32_43_049: [ If handle is NULL, file_destroy shall return. ]*/
+TEST_FUNCTION(file_destroy_called_with_null_handle)
+{
+    ///arrange
+
+    ///act
+    file_destroy(NULL);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+}
+
+/*Tests_SRS_FILE_WIN32_43_011: [ file_destroy shall wait for all I / O to complete by calling WaitForThreadpoolIoCallbacks.]*/
+/*Tests_SRS_FILE_WIN32_43_012: [ file_destroy shall close the cleanup group by calling CloseThreadpoolCleanupGroup.]*/
+/*Tests_SRS_FILE_WIN32_43_013: [ file_destroy shall destroy the environment by calling DestroyThreadpoolEnvironment.]*/
+/*Tests_SRS_FILE_WIN32_43_016: [ file_destroy shall call CloseHandle on handle->h_file]*/
+/*Tests_SRS_FILE_WIN32_43_015: [ file_destroy shall close the threadpool IO by calling CloseThreadPoolIo.]*/
+/*Tests_SRS_FILE_WIN32_43_042: [ file_destroy shall free the handle.]*/
+TEST_FUNCTION(file_destroy_succeeds)
+{
+
+    PTP_WIN32_IO_CALLBACK captured_callback;
+    FILE_HANDLE file_handle = get_file_handle("file_destroy_succeeds.txt", &captured_callback);
+    PTP_IO captured_ptp_io;
+
+    STRICT_EXPECTED_CALL(mock_WaitForThreadpoolIoCallbacks(IGNORED_ARG, FALSE))
+        .CaptureArgumentValue_pio(&captured_ptp_io);
+    STRICT_EXPECTED_CALL(mock_CloseThreadpoolCleanupGroup(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(mock_DestroyThreadpoolEnvironment(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(mock_CloseHandle(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(mock_CloseThreadpoolIo(IGNORED_ARG))
+        .ValidateArgumentValue_pio(&captured_ptp_io);
+    STRICT_EXPECTED_CALL(free(file_handle));
+
+    ///act
+    file_destroy(file_handle);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+}
+
+/*Tests_SRS_FILE_43_009: [ If handle is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+/*Tests_SRS_FILE_WIN32_43_043: [ If handle is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+TEST_FUNCTION(file_write_async_fails_with_null_handle)
+{
+    ///arrange
+    unsigned char source[10];
+    FILE_CB user_callback = (FILE_CB)42;
+
+    ///act
+    FILE_WRITE_ASYNC_RESULT result = file_write_async(NULL, source, 1, 1, user_callback, NULL);
+
+    ///assert
+    ASSERT_ARE_EQUAL(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_INVALID_ARGS, result);
+}
+
+/*Tests_SRS_FILE_43_010: [ If source is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+/*Tests_SRS_FILE_WIN32_43_044: [ If source is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+TEST_FUNCTION(file_write_async_fails_with_null_source)
+{
+    ///arrange
+    FILE_HANDLE file_handle = (FILE_HANDLE)42;

```

Comment:
>FILE_HANDLE [](start = 31, length = 11)

Should always pass a real handle that was created (except the NULL test)

Reply:
Here and everywhere

---
In reply to: [456719743](https://github.com/Azure/azure-c-pal/pull/10#discussion_r456719743) [](ancestors = 456719743)